ql/src/java/org/apache/hadoop/hive/llap/ProactiveEviction.java

```diff
@@ -0,0 +1,311 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.llap;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import javax.net.SocketFactory;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.common.io.CacheTag;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.llap.daemon.rpc.LlapDaemonProtocolProtos;
+import org.apache.hadoop.hive.llap.impl.LlapManagementProtocolClientImpl;
+import org.apache.hadoop.hive.llap.registry.LlapServiceInstance;
+import org.apache.hadoop.hive.llap.registry.impl.LlapRegistryService;
+import org.apache.hadoop.io.retry.RetryPolicies;
+import org.apache.hadoop.io.retry.RetryPolicy;
+import org.apache.hadoop.net.NetUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Through this class the caller (typically HS2) can request eviction of buffers from LLAP cache by specifying a DB,
+ * table or partition name/(value). Request sending is implemented here.
+ */
+public final class ProactiveEviction {
+
+  private ProactiveEviction() {
+    // Not to be used;
+  }
+
+  /**
+   * Trigger LLAP cache eviction of buffers related to entities residing in request parameter.
+   * @param conf
+   * @param request
+   */
+  public static void evict(Configuration conf, Request request) {
+    if (!HiveConf.getBoolVar(conf, HiveConf.ConfVars.LLAP_IO_PROACTIVE_EVICTION_ENABLED)) {
+      return;
+    }
+
+    try {
+      LlapRegistryService llapRegistryService = LlapRegistryService.getClient(conf);
+      Collection<LlapServiceInstance> instances = llapRegistryService.getInstances().getAll();
+      if (instances.size() == 0) {
+        // Not in LLAP mode.
+        return;
+      }
+      ExecutorService executorService = Executors.newCachedThreadPool();
```

Comment:
This code has multiple issues and it is not safe.

1. First issue, there is no explicit shutdown of the executor and that is a big issue by it self,  because some implementations shutdown themselves in a finalizer **BUT JVM spec do not guaranteed to ever be call finalize()**.
2. Second this code is creating a thread pool on each call, that is a huge resource wast.
3. Third point Threads are not named thus makes it hard to debug.
4. Fourth assume some of those tasks hangs on IO what will happen ? will the JVM ever stops ?
5. Nth issue Assume assume we have 500 llap nodes (yes we have cluster with 500 node running), this will fire 500 threads is that okay ?

IMO First cut of this pr should avoid all this Async stuff and start with a simple version OR you need to have a static fixed thread pool with a well defined **life cycle** and clear strategy on how to deal with **blocking IO**, **errors** and **shutdown**

Reply:
To be more clear about point one you can read the [java spec](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html) it is always very good point to start **An unused ExecutorService should be shut down to allow reclamation of its resources.**

Reply:
For number 4 you can refer to this spec https://wiki.sei.cmu.edu/confluence/display/java/TPS02-J.+Ensure+that+tasks+submitted+to+a+thread+pool+are+interruptible

Reply:
I hope you are convinced that Asynchronous stuff can be trick to get right and most of the benefit can turn into debugging nightmare. Let me know if you have more questions

Reply:
Thanks for the detailed comments! I agree with all the points above, I did miss the fact that I was creating a new pool at each invocation - my bad.
I've amended the change with https://github.com/szlta/hive/commit/74364b022f14edfa597a7d95b0774def8f3ddefd to fix this and address the other points as well.

Reply:
@szlta the new commit introduced new problem, The problem is that this does not work when using multiple threads each thread will see a null executor and each one will create a new one and same with the shutdown method. 
One way to fix this is to make the executor static and final, therefore will be constructed at the time class loading. Also shutdown hook might be needed here.

Reply:
You can check this link about Double checking locking https://en.wikipedia.org/wiki/Double-checked_locking

Reply:
Makes sense. I guess we can do something like this then: https://github.com/apache/hive/pull/909/commits/4b285ef48e95f0a7b83029ab42de7a535d93d4d5

Reply:
Hi @b-slim can you take a look on the recent commit please?